utxo/src/utxo_impl/mod.rs

```diff
@@ -0,0 +1,537 @@
+//TODO: remove once the functions are used.
+#![allow(dead_code)]
+use crate::{Error, TxUndo};
+use common::chain::{OutPoint, OutPointSourceId, Transaction, TxOutput};
+use common::primitives::{BlockHeight, Id, Idable};
+use std::collections::BTreeMap;
+use std::fmt::{Debug, Formatter};
+
+use common::chain::block::Block;
+use parity_scale_codec::{Decode, Encode};
+
+pub mod utxo_storage;
+
+//todo: proper placement and derivation of this max
+const MAX_OUTPUTS_PER_BLOCK: u32 = 500;
+
+// Determines whether the utxo is for the blockchain of for mempool
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub enum UtxoSource {
+    /// At which height this containing tx was included in the active block chain
+    BlockChain(BlockHeight),
+    MemPool,
+}
+
+impl UtxoSource {
+    fn is_mempool(&self) -> bool {
+        match self {
+            UtxoSource::BlockChain(_) => false,
+            UtxoSource::MemPool => true,
+        }
+    }
+
+    fn blockchain_height(&self) -> Result<BlockHeight, Error> {
+        match self {
+            UtxoSource::BlockChain(h) => Ok(*h),
+            UtxoSource::MemPool => Err(crate::Error::NoBlockchainHeightFound),
+        }
+    }
+}
+
+/// The Unspent Transaction Output
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct Utxo {
+    output: TxOutput,
+    is_block_reward: bool,
+    /// identifies whether the utxo is for the blockchain or for mempool.
+    source: UtxoSource,
+}
+
+impl Utxo {
+    pub fn new(output: TxOutput, is_block_reward: bool, height: BlockHeight) -> Self {
+        Self {
+            output,
+            is_block_reward,
+            source: UtxoSource::BlockChain(height),
+        }
+    }
+
+    /// a utxo for mempool, that does not need the block height.
+    pub fn new_for_mempool(output: TxOutput, is_block_reward: bool) -> Self {
+        Self {
+            output,
+            is_block_reward,
+            source: UtxoSource::MemPool,
+        }
+    }
+
+    pub fn is_block_reward(&self) -> bool {
+        self.is_block_reward
+    }
+
+    pub fn height(&self) -> &UtxoSource {
+        &self.source
+    }
```

Comment:
I changed it to source_height. I think that's OK. I want something to indicate that this is about the height.

Option is not OK because it's not clear what None really means.

Reply:
Resolve if you're happy with this.